test/middleware-test.ts

```diff
@@ -0,0 +1,696 @@
+// /test/middleware-test.ts
+// Tests the rate limiting middleware
+
+import { jest } from '@jest/globals'
+import Express from 'express'
+import request from 'supertest'
+
+import rateLimit, {
+	Store,
+	Options,
+	IncrementResponse,
+} from '../source/index.js'
+
+import { createServer } from './helpers/create-server.js'
+
+describe('middleware test', () => {
+	afterEach(() => {
+		jest.restoreAllMocks()
+	})
+
+	class MockStore implements Store {
+		initWasCalled = false
+		incrementWasCalled = false
+		decrementWasCalled = false
+		resetKeyWasCalled = false
+
+		counter = 0
+
+		init(_options: Options): void {
+			this.initWasCalled = true
+		}
+
+		async increment(_key: string): Promise<IncrementResponse> {
+			this.counter += 1
+			this.incrementWasCalled = true
+
+			return { totalHits: this.counter, resetTime: undefined }
+		}
+
+		async decrement(_key: string): Promise<void> {
+			this.counter -= 1
+			this.decrementWasCalled = true
+		}
+
+		async resetKey(_key: string): Promise<void> {
+			this.resetKeyWasCalled = true
+		}
+	}
+
+	it('should not modify the options object passed', () => {
+		const options = {}
+		rateLimit(options)
+		expect(options).toStrictEqual({})
+	})
+
+	it('should call `init` even if no requests have come in', async () => {
+		const store = new MockStore()
+		rateLimit({
+			store,
+		})
+
+		expect(store.initWasCalled).toEqual(true)
+	})
+
+	it('should print an error if `request.ip` is undefined', async () => {
+		jest.spyOn(global.console, 'error').mockImplementation(() => {})
+
+		await Promise.resolve(
+			rateLimit()(
+				{ ip: undefined } as any as Express.Request,
```

Comment:
I didn't know you could do `as` twice in a row. Cool.

Reply:
It's usually used when Typescript puts it foot down and says 'you may not!' 😆

Reply:
I usually just suppress the errors with a `@ts-expect-error` comment if its a test though